### PR TITLE
tcp_peer: Fix possible goroutine leak

### PR DESCRIPTION
**Pull request by stepanovdmitrii**
### Problem

Scenario:
1. Two messages were read from the connection `p.conn`
2. The first message has started to be processed
3. The second message was queued to be added to the channel `p.incoming`
4. Processing of the first message failed with an error
5. TCP peer is closed, but processing of the second message continues

### Solution

Add `p.done` channel close handling.

### CODE REVIEW

File: pkg/network/tcp_peer.go [hunk 1]
```diff
@@ -164,6 +164,7 @@ func (p *TCPPeer) handleConn() {
 	err = p.SendVersion()
 	if err == nil {
 		r := io.NewBinReaderFromIO(p.conn)
+		var closed bool
 		for {
```
Comment (roman-khimov) [inline pkg/network/tcp_peer.go:168]: You can use a label here...

File: pkg/network/tcp_peer.go [hunk 2]
```diff
@@ -174,7 +175,14 @@ func (p *TCPPeer) handleConn() {
 			} else if err != nil {
 				break
 			}
-			p.incoming <- msg
+			select {
+			case p.incoming <- msg:
+			case <-p.done:
+				closed = true
```
Comment (roman-khimov) [inline pkg/network/tcp_peer.go:181]: ...and `break lbl` here.

File: pkg/network/tcp_peer.go [hunk 3]
```diff
@@ -174,7 +175,14 @@ func (p *TCPPeer) handleConn() {
 			} else if err != nil {
 				break
 			}
-			p.incoming <- msg
+			select {
+			case p.incoming <- msg:
+			case <-p.done:
```
Comment (roman-khimov) [inline pkg/network/tcp_peer.go:180]: This needs to be the first in `select` then.

Comment (stepanovdmitrii) [reply pkg/network/tcp_peer.go:180]: Why?

File: pkg/network/tcp_peer.go (hunk 2 above)
Comment (stepanovdmitrii) [reply pkg/network/tcp_peer.go]: Oh, ok, fixed.

File: pkg/network/tcp_peer.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/network/tcp_peer.go:180]: Sorry, I was thinking about something else.

**Review by roman-khimov — APPROVED** [review]
Yeah, looks legit. Likely the problem was introduced with f78bd6474f2d0131b0750959e9c3bd80d45c8e66.

**Review by AnnaShaleva — APPROVED** [review]

**Comment by AnnaShaleva** [comment]
Nice catch, waiting for the testing jobs to finish.

**Pull request merged**